Avoid IPC to OpenDirectory XPC service on main thread

pvollan opened the pull request:
#### 1a7bea21fb470e352743786b98cac1b5b5e6ca5a
<pre>
Avoid IPC to OpenDirectory XPC service on main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=272192">https://bugs.webkit.org/show_bug.cgi?id=272192</a>
<a href="https://rdar.apple.com/123778399">rdar://123778399</a>

Reviewed by Sihui Liu.

Avoid IPC to OpenDirectory XPC service on main thread, since it can potentially block the main thread for a long time.
This IPC is called in response to a notification from the UI process when the Open Directory cache has been invalidated.
There is a window of time where calling getpwuid_r and related functions in the WebContent process can return stale
values after the cache has been invalidated. This is still an issue after this patch and should be addressed, but it
may be outside the scope of this patch. WebKit is only calling this function before entering the sandbox, where we will
never get stale values. It is however possible that other frameworks are calling this function, but I am not aware of
that being the case, currently.

* Source/WebKit/GPUProcess/mac/GPUProcessMac.mm:
(WebKit::GPUProcess::openDirectoryCacheInvalidated):
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::AuxiliaryProcess::openDirectoryCacheInvalidated):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::openDirectoryCacheInvalidated):

Canonical link: <a href="https://commits.webkit.org/277204@main">https://commits.webkit.org/277204@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fbff4a15af91ffa44bd69cd0e02e7732c6029c86

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46829 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25990 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49452 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49507 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42877 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30400 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23456 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38139 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47409 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22991 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40326 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19441 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20318 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4875 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43060 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41832 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51380 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21839 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18179 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45429 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23126 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44395 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23636 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6588 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22836 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/GPUProcess/mac/GPUProcessMac.mm:
```diff
@@ -111,10 +112,12 @@
 {
-    AuxiliaryProcess::openDirectoryCacheInvalidated(WTFMove(handle));
+    auto cacheInvalidationHandler = [handle = WTFMove(handle)] () mutable {
+        AuxiliaryProcess::openDirectoryCacheInvalidated(WTFMove(handle));
```
My understanding is this function populates the cache, how do we make sure the cache will not be used before the function is called, after this patch delays it?

pvollan replied:
That is a very good point. We actually subscribe to a notification about this in the UI process, and then send a message to the GPU and WebContent process to give them a chance to repopulate the cache. Perhaps dispatching the work to a non main thread will not add much to the overall time it takes to populate the cache? The information in the cache is needed when calling `getpwuid_r`, which we only call on startup from `getHomeDirectory()` in WebKit before we are entering the sandbox. However, there could possibly be other frameworks that are calling this function or other functions related to the home directory of the user after entering the sandbox.

Thanks for reviewing!

pvollan replied:
I expanded the commit message to explain this closer.

pvollan committed:
Avoid IPC to OpenDirectory XPC service on main thread
https://bugs.webkit.org/show_bug.cgi?id=272192
rdar://123778399

Reviewed by Sihui Liu.

Avoid IPC to OpenDirectory XPC service on main thread, since it can potentially block the main thread for a long time.
This IPC is called in response to a notification from the UI process when the Open Directory cache has been invalidated.
There is a window of time where calling getpwuid_r and related functions in the WebContent process can return stale
values after the cache has been invalidated. This is still an issue after this patch and should be addressed, but it
may be outside the scope of this patch. WebKit is only calling this function before entering the sandbox, where we will
never get stale values. It is however possible that other frameworks are calling this function, but I am not aware of
that being the case, currently.

* Source/WebKit/GPUProcess/mac/GPUProcessMac.mm:
(WebKit::GPUProcess::openDirectoryCacheInvalidated):
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::AuxiliaryProcess::openDirectoryCacheInvalidated):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::openDirectoryCacheInvalidated):

Canonical link: https://commits.webkit.org/277204@main
